pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/SanitizationTransformer.java

```diff
@@ -32,51 +34,197 @@
+  private static final String NULL_CHARACTER = "\0";
+  private final Map<String, SanitizedColumnInfo> _columnToColumnInfoMap = new HashMap<>();
 
   public SanitizationTransformer(Schema schema) {
+    FieldSpec.MaxLengthExceedStrategy maxLengthExceedStrategy;
     for (FieldSpec fieldSpec : schema.getAllFieldSpecs()) {
-      if (!fieldSpec.isVirtualColumn() && fieldSpec.getDataType() == DataType.STRING) {
-        _stringColumnMaxLengthMap.put(fieldSpec.getName(), fieldSpec.getMaxLength());
+      if (!fieldSpec.isVirtualColumn()) {
+        switch (fieldSpec.getDataType()) {
+          case STRING:
+            maxLengthExceedStrategy = fieldSpec.getMaxLengthExceedStrategy() == null
+                ? FieldSpec.MaxLengthExceedStrategy.TRIM_LENGTH : fieldSpec.getMaxLengthExceedStrategy();
+            _columnToColumnInfoMap.put(fieldSpec.getName(), new SanitizedColumnInfo(fieldSpec.getName(),
+                fieldSpec.getMaxLength(), maxLengthExceedStrategy, fieldSpec.getDefaultNullValueString()));
+            break;
+          case JSON:
+          case BYTES:
+            maxLengthExceedStrategy = fieldSpec.getMaxLengthExceedStrategy() == null
+                ? FieldSpec.MaxLengthExceedStrategy.NO_ACTION : fieldSpec.getMaxLengthExceedStrategy();
+            _columnToColumnInfoMap.put(fieldSpec.getName(), new SanitizedColumnInfo(fieldSpec.getName(),
+                fieldSpec.getMaxLength(), maxLengthExceedStrategy, fieldSpec.getDefaultNullValueString()));
+            break;
+          default:
+            // Do nothing for other data types
+        }
       }
     }
   }
 
   @Override
   public boolean isNoOp() {
-    return _stringColumnMaxLengthMap.isEmpty();
+    return _columnToColumnInfoMap.isEmpty();
   }
 
   @Override
   public GenericRow transform(GenericRow record) {
-    for (Map.Entry<String, Integer> entry : _stringColumnMaxLengthMap.entrySet()) {
-      String stringColumn = entry.getKey();
-      int maxLength = entry.getValue();
-      Object value = record.getValue(stringColumn);
-      if (value instanceof String) {
-        // Single-valued column
-        String stringValue = (String) value;
-        String sanitizedValue = StringUtil.sanitizeStringValue(stringValue, maxLength);
-        // NOTE: reference comparison
-        //noinspection StringEquality
-        if (sanitizedValue != stringValue) {
-          record.putValue(stringColumn, sanitizedValue);
+    for (Map.Entry<String, SanitizedColumnInfo> entry : _columnToColumnInfoMap.entrySet()) {
+      String columnName = entry.getKey();
+      Object value = record.getValue(columnName);
+      Pair<?, Boolean> result;
+      if (value instanceof byte[]) {
+        // Single-values BYTES column
+        result = sanitizeBytesValue(columnName, (byte[]) value, entry.getValue());
+        record.putValue(columnName, result.getLeft());
+        if (result.getRight()) {
+          record.putValue(GenericRow.INCOMPLETE_RECORD_KEY, true);
```

Comment:
We don't want to put `INCOMPLETE_RECORD_KEY` because it is used for error case when `continueOnError` is enabled.
If the intention is to track the values modified, we might need to add a different key. Suggest adding that in a separate PR

Reply:
It was similar effort to remove this so i added `SANITIZED_RECORD_KEY` instead